### ISSUE TITLE
Before Tax After Tax Register Adjustments

### ISSUE THREAD

**Issue by smockensturm** [verbb/gift-voucher#78]
https://github.com/verbb/gift-voucher/blob/a7f284f86e229599510f61cedbadfb1d5152de70/src/GiftVoucher.php#L251

Did you forget something?

     `private function _registerAdjusters()
    {
        Event::on(OrderAdjustments::class, OrderAdjustments::EVENT_REGISTER_ORDER_ADJUSTERS, function(RegisterComponentTypesEvent $event) {
            $settings = $this->getSettings();

            // Re-order the built-in adjusters to ensure gift vouchers are applied before tax.
            if ($settings->registerAdjuster === 'beforeTax') {
                $types = $event->types;

                // Find the Tax adjuster, it should go before that, but if its not found (Commerce Lite), append
                $taxKey = array_search(Tax::class, $event->types);

                if ($taxKey) {
                    array_splice($types, $taxKey, 0, GiftVoucherAdjuster::class);
                } else {
                    $types[] = GiftVoucherAdjuster::class;
                }

                $event->types = $types;
            } elseif ($settings->registerAdjuster === 'afterTax') {
                $types[] = GiftVoucherAdjuster::class;
            }
        });
    }`

**Comment by engram-design** [comment]
You shouldn't need to apply the adjuster after tax anymore https://github.com/verbb/gift-voucher/issues/69#issuecomment-894791392, but I've added this back in for you.

Fixed for the next release. To get the fix early, change your `verbb/gift-voucher` requirement in `composer.json` to:

```
"require": {
  "verbb/gift-voucher": "dev-craft-3 as 2.6.3",
  "...": "..."
}
```

Then run `composer update`.

**Comment by smockensturm** [comment]
What a mess. Thanks.

**Comment by smockensturm** [comment]
- `registerAdjuster` - Controls when the adjuster should be applied. Valid options are `beforeTax` (default) and `afterTax`.

Probably ought to take that out then.

**Comment by engram-design** [comment]
I'll keep it in for the moment, there's been a fair bit of back-and-forth on this issue (as you can tell), but no massive harm in keeping it for now.

**Comment by smockensturm** [comment]
at least add a caveat that this issue is not settled. would have saved me some time. your docs are making a statement that just isn’t true.

**Comment by engram-design** [comment]
This issue should be settled - but you're right that the docs hadn't made this change, simply out of date.

As you mention in https://github.com/verbb/gift-voucher/issues/69#issuecomment-965822854 are you having issues with the discount not working as expected? If you can provide some details in this thread, happy to investigate.

Having this adjusters registered `beforeTax` should still be the default. It'll discount the line item price as well as discount the shipping. If both these values are $0, then there's no tax on $0. If the voucher is partially redeemed, the correct tax for the discount price will be correct.

There's only a specific case where you'd want to use `afterTax` where you actually **want** the tax still applied for shipping, which I believe might be due to VAT. One such customer requested this due to their needs, despite the shipping total being $0, they still wanted shipping tax to be charged.

**Comment by smockensturm** [comment]
again, great intel. shall report back. thank you.

**Comment by smockensturm** [comment]
Also... just sent an admin invite. It's all US tho so you'll have to pretend you live in Ohio when you check out. :)

**Comment by engram-design** [comment]
Thanks @smockensturm will take a look

**Comment by engram-design** [comment]
Okay, so in this case, you'll want to swap the discount to be applied `afterTax` which should have the desired effect.

![image](https://user-images.githubusercontent.com/1221575/141610795-7271e8e5-011f-4976-ba6f-0e56d5f659ed.png)

You'll notice it'll still show the amount in the order totals, but it won't need to be payed. This is exactly the situation we're taking care of by using the additional `GiftVoucherShippingAdjuster` adjuster, to take off the amount for shipping, so it doesn't appear.

**Comment by smockensturm** [comment]
Hey did you put this back in or not?

Our client needs the tax to be applied to the voucher. See my comment [over here](https://github.com/verbb/gift-voucher/issues/69#issuecomment-965822854)

And the `beforeTax` / `afterTax` config appears to do nothing.

**Comment by engram-design** [comment]
It's been removed due to it being unneeded now (changes in https://github.com/verbb/gift-voucher/issues/69), but will follow up in https://github.com/verbb/gift-voucher/issues/69#issuecomment-986792330 as I know there's more to discuss here.

**Comment by engram-design** [comment]
Fixed in [2.7.0](https://github.com/verbb/gift-voucher/releases/tag/2.7.0)
